src/Storages/MergeTree/MergeTreeReaderCompact.cpp

```diff
@@ -239,4 +240,53 @@ bool MergeTreeReaderCompact::isContinuousReading(size_t mark, size_t column_posi
         || (mark == last_mark + 1 && column_position == 0 && last_column == data_part->getColumns().size() - 1);
 }
 
+
+size_t MergeTreeReaderCompact::getReadBufferSize()
```

Comment:
I think I've got the idea: we need for find the largest granule and limit buffer size if possible.
Still, comment will be helpful: what this function does and what why do we care calculating this stuff instead of using `max_read_buffer_size`.

Also, it is not clear what would we use inside function. Like we need `column_positions` filled, `data_part`, `marks_loader` (is it all?). This function may be made static with implicit arguments to avoid possible usage of uninitialized members.